CMakeLists.txt

```diff
@@ -1,53 +1,118 @@
+      Debug
+      MinSizeRel
+      None
+      RelWithDebInfo
+      Release
+      Coverage
+      Profile)
+  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "${allowableBuildTypes}")
+  if(NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE Debug CACHE STRING "" FORCE)
+  elseif(NOT CMAKE_BUILD_TYPE IN_LIST allowableBuildTypes)
+    message(FATAL_ERROR "Invalid build type : ${CMAKE_BUILD_TYPE}")
+  endif()
+endif()
+
+if(POLICY CMP0048)
+  cmake_policy(SET CMP0048 NEW)
+endif()
+
+if(POLICY CMP0083)
+  cmake_policy(SET CMP0083 NEW)
+  include(CheckPIESupported)
+  check_pie_supported(OUTPUT_VARIABLE output LANGUAGES CXX)
+  if(NOT CMAKE_CXX_LINK_PIE_SUPPORTED)
+    message(WARNING "PIE is not supported at link time: ${output}.\n" "PIE link options will not be passed to linker.")
+  else()
+    set(CMAKE_POSITION_INDEPENDENT_CODE TRUE)
+    message(STATUS "PIE link options will be passed to linker.")
+  endif()
+else()
+  set(CMAKE_POSITION_INDEPENDENT_CODE TRUE)
+endif()
+
+# Optional IPO. Do not use IPO if it's not supported by compiler.
+include(CheckIPOSupported)
+check_ipo_supported(RESULT result OUTPUT output)
+if(result)
+  set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)
+else()
+  message(WARNING "IPO is not supported: ${output}")
+endif()
+
+option(ENABLE_ALL_WARNINGS "Enable all warnings" TRUE)
+
+include(CompilerOptions)
 
 #----------------------------------------------------------------------------
 # Find Geant4 package, activating all available UI and Vis drivers by default
 # You can set WITH_GEANT4_UIVIS to OFF via the command line or ccmake/cmake-gui
 # to build a batch mode only executable
 #
 
-option(WITH_GEANT4_UIVIS "Build example with Geant4 UI and Vis drivers" ON)
-if(WITH_GEANT4_UIVIS)
-  find_package(Geant4 REQUIRED ui_all vis_all)
+option(WITH_GEANT4_VIS "Build example with Geant4 Vis drivers" ON)
+if(WITH_GEANT4_VIS)
+  find_package(Geant4 REQUIRED vis_all)
 else()
   find_package(Geant4 REQUIRED)
 endif()
 
-find_package( ROOT REQUIRED )
-find_package( LCIO 2.11 REQUIRED )
-
-FOREACH( pkg LCIO ROOT Geant4 )
-	IF( ${pkg}_FOUND )
-		INCLUDE_DIRECTORIES( SYSTEM ${${pkg}_INCLUDE_DIRS} )
-		LINK_LIBRARIES( ${${pkg}_LIBRARIES} )
-		ADD_DEFINITIONS ( ${${pkg}_DEFINITIONS} )
-	ENDIF()
-ENDFOREACH()
-
-
-ADD_DEFINITIONS( "-Wall -ansi -pedantic -std=c++11 -Weffc++" )
-
-include( ${Geant4_USE_FILE} )
-include_directories( ${PROJECT_SOURCE_DIR}/include )
-
-file(GLOB sources ${PROJECT_SOURCE_DIR}/src/*.cpp)
+find_package(ROOT REQUIRED)
+find_package(LCIO 2.11 REQUIRED)
 
-file(GLOB headers ${PROJECT_SOURCE_DIR}/include/*.h ${PROJECT_SOURCE_DIR}/include/*.hpp)
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED OFF)
+set(CMAKE_CXX_EXTENSIONS OFF)
 
-add_library( MyLib ${sources} ${headers} )
+file(DOWNLOAD "https://github.com/nlohmann/json/releases/download/v3.7.3/json.hpp" "${CMAKE_INSTALL_PREFIX}/include/json.hpp" SHOW_PROGRESS EXPECTED_HASH SHA256=3b5d2b8f8282b80557091514d8ab97e27f9574336c804ee666fda673a9b59926)
```

Comment:
Don't find it necessarily good. 
Will it be possible to rebuild when working 
offline ?

Reply:
If you first run online yes.... It check the sha256sum and if  it matches it doesn't download it again

Reply:
So it should be quite safe because if you pushed it you are online. The only draw back is  : 
1) You clone go offline and try to run cmake.
2) Erase bin folder go online and run cmake again.
I prefer this than polluting the code A with a code B more than 10 times bigger and it's very easy to check if new version works etc

Reply:
I'm trying to install the software on a centos 8 machine and I'm facing the following problem : 

> CMake Error at CMakeLists.txt:114 (file):
  file DOWNLOAD HASH mismatch

    for file: [/tmp/toErase/SDHCALSim/bin/include/json.hpp]
      expected hash: [3b5d2b8f8282b80557091514d8ab97e27f9574336c804ee666fda673a9b59926]
        actual hash: [e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855]
             status: [1;"Unsupported protocol"]


I had already the same issue for getting GEANT4 data. Probably there is  an issue with the centos8 machine. But it would be good to have a workaround for this kind of issue.

Reply:
In fact, the file downloaded by cmake has zero size. 
 By using another cmake, I've succeeded. 
Possible explanation is that when I've installed ilcsoft, openSSL was not installed and this is needed to download files by cmake. 
It would be good to add to the cmake files something explaining to the user how to get around the problem when cmake fails.
Something like a message saying download yourself the file, put it there and rerun cmake.

Reply:
Yes i will try to find a workaround... It's strange openssl is not installed by your operating system. I don't know if you can pass option to the python script to download programs in ilcsoft. CMake can be compiled with is own openssl i think.

Reply:
I can download it and put it on a named  "external" folder to avoid downloading it. It's a pitty not to be able to use great features :(

Reply:
CMake Error at CMakeLists.txt:115 (file):
  file DOWNLOAD HASH mismatch

    for file: [/home/rpclab/Desktop/SDHCALSim3/bin/include/json.hpp]
      expected hash: [3b5d2b8f8282b80557091514d8ab97e27f9574336c804ee666fda673a9b59926]
        actual hash: [e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855]
             status: [22;"HTTP response code said error"]



CMake Warning at CMakeLists.txt:119 (message):
  ERROR : "HTTP response code said error"

  CMake can't download the file
  "https://github.com/nlohmann/json/releases/download/v3.7.3/json.hpp" ! Do
  it by yourself and add it to
  "/home/rpclab/Desktop/SDHCALSim3/bin/include/json.hpp" and rerun CMake !

This kind of help is ok for you?